app/views/core/CreateAccountModal/BasicInfoView.coffee

```diff
@@ -0,0 +1,249 @@
+ModalView = require 'views/core/ModalView'
+AuthModal = require 'views/core/AuthModal'
+template = require 'templates/core/create-account-modal/basic-info-view'
+forms = require 'core/forms'
+errors = require 'core/errors'
+User = require 'models/User'
+State = require 'models/State'
+
+###
+This view handles the primary form for user details — name, email, password, etc,
+and the AJAX that actually creates the user.
+
+It also handles facebook/g+ login, which if used, open one of two other screens:
+sso-already-exists: If the facebook/g+ connection is already associated with a user, they're given a log in button
+sso-confirm: If this is a new facebook/g+ connection, ask for a username, then allow creation of a user
+
+The sso-confirm view *inherits from this view* in order to share its account-creation logic and events.
+This means the selectors used in these events must work in both templates.
+
+This view currently uses the old form API instead of stateful render.
+It needs some work to make error UX and rendering better, but is functional.
+###
+
+module.exports = class BasicInfoView extends ModalView
+  id: 'basic-info-view'
+  template: template
+
+  events:
+    'input input[name="name"]': 'onInputName'
+    'click .back-button': 'onClickBackButton'
+    'submit form': 'onSubmitForm'
+    'click .login-link': 'onClickLoginLink'
+    'click .use-suggested-name-link': 'onClickUseSuggestedNameLink'
+    'click #facebook-signup-btn': 'onClickSsoSignupButton'
+    'click #gplus-signup-btn': 'onClickSsoSignupButton'
+
+  initialize: ({ @sharedState } = {}) ->
+    @state = new State {
+      suggestedName: null
+    }
+    @onNameChange = _.debounce(_.bind(@checkNameUnique, @), 500)
+    @listenTo @sharedState, 'change:facebookEnabled', -> @renderSelectors('.auth-network-logins')
+    @listenTo @sharedState, 'change:gplusEnabled', -> @renderSelectors('.auth-network-logins')
+
+  checkNameUnique: ->
+    name = $('input[name="name"]', @$el).val()
+    return forms.clearFormAlerts(@$('input[name="name"]').closest('.form-group').parent()) if name is ''
+    @nameUniquePromise = new Promise((resolve, reject) => User.getUnconflictedName name, (newName) =>
+      if name is newName
+        @state.set { suggestedName: null }
+        @clearNameError()
+        resolve true
+      else
+        @state.set { suggestedName: newName }
+        @setNameError(newName)
+        resolve false
+    )
```

Comment:
I think the issue is that the Promise should be 1) single-purpose and 2) without (local) side-effects. All the Promise should do is check if the name conflicts and report if it does or doesn't, not doing anything to the view. Perhaps this would be made better with memoize...

```coffeescript
  initialize: ->
    @checkNameDebounced = _.debounce(_.bind(@checkNameUnique, @), 500)

  onNameChange: ->
    @clearNameConflictAlerts() # removes either success or failure alerts instantly, no wait
    @checkNameDebounced()

  checkNameDebounced: ->
    name = $('input[name="name"]', @$el).val()
    @checkNameUnique(name).then (result) =>
      currentName = $('input[name="name"]', @$el).val()
      return if result.name isnt currentName # name has changed since check began, so ignore. Later debounced onNameChange will check the latest value
      @showNameConflictAlerts(result) # shows either success or failure alerts

  onFormSubmit: (e) ->
    # disable the form, do other checks, show loading here...
    name = $('input[name="name"]', @$el).val()
    checkNameUnique(name).then (result) ->
      if (result.conflicts)
        @showNameConflictAlerts(result)
      else
        @makeUser() # or whatever is the next step in the process

  checkNameUnique: _.memoize (name) ->
    return new Promise((resolve, reject) => User.getUnconflictedName name, (newName) =>
      resolve {
        name: name
        newName: newName
        conflicts: name isnt newName
      }
```

This way, whenever you call `checkNameUnique`, you get a unique promise for that name. It doesn't matter if the request to check that name is finished, in progress, or unstarted, the result will be a promise that will asynchronously return the result. Then the event handlers for input change or form submission can each have their own unique handling (the form moves the process along on success, changing the input does not).